### PR TITLE
Add CollectAndDistributeFpnRpnProposalsOp for FPN support

### PR DESCRIPTION
**Pull request by orionr**
* Adds a C++ operator equivalent to the Python op in Detectron
* Once some additional GenerateProposalsOp changes are made this will
 let us support Detectron FPN models with straight Caffe2 C++ ops
* RetinaNet and segmentation models require additional work

### CODE REVIEW

File: caffe2/operators/collect_and_distribute_fpn_rpn_proposals_op.cc [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+// Compute the area of an array of boxes.
+ERArrXXf BoxesArea(const ERArrXXf& boxes) {
+  // equivalent to python code
+  //   w = (boxes[:, 2] - boxes[:, 0] + 1)
+  //   h = (boxes[:, 3] - boxes[:, 1] + 1)
+  //   areas = w * h
+  //   assert np.all(areas >= 0), 'Negative areas founds'
+  const auto w = boxes.col(2) - boxes.col(0) + 1;
+  const auto h = boxes.col(3) - boxes.col(1) + 1;
+  const auto areas = w * h;
```
[the remark this answers is not in the document]
Comment (orionr) [reply caffe2/operators/collect_and_distribute_fpn_rpn_proposals_op.cc]: Good catch.

**Comment by orionr** [comment]
@newstzpz thanks for the review. Things should be addressed - let me know if you see anything else.

**Comment by orionr** [comment]
Actually - need to try and remove the use of `conservativeResize` as well. Stay tuned.

**Comment by orionr** [comment]
And should now be ready for review. Thanks!

File: caffe2/operators/collect_and_distribute_fpn_rpn_proposals_op.cc [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+REGISTER_MKL_OPERATOR(
+    CollectAndDistributeFpnRpnProposals,
+    mkl::MKLFallbackOp<CollectAndDistributeFpnRpnProposalsOp<CPUContext>>);
+#endif // CAFFE2_HAS_MKL_DNN
+
+OPERATOR_SCHEMA(CollectAndDistributeFpnRpnProposals)
+    .NumInputs(2, INT_MAX)
+    .NumOutputs(3, INT_MAX)
+    .SetDoc(R"DOC(
+Merge RPN proposals generated at multiple FPN levels and then
```
Comment (rbgirshick) [inline caffe2/operators/collect_and_distribute_fpn_rpn_proposals_op.cc:262]: Maybe add two notes to the op documentation:
1) Only inference is supported (the Python version has a training mode in which it dynamically builds a training minibatch for the detection head(s) based on the FPN RPN proposals)
2) The inputs documented below are only an example; if the min/max levels are changed, then the number of inputs and outputs, as well as their level numbering, will change

**Review by rbgirshick — APPROVED** [review]
The code looks good! One minor suggestion for the op documentation.

**Pull request merged**